llext: fix link order for weak syscall symbols

pillo79 opened the pull request:
This is an alternative to #79023. Instead of the generic machinery added there, this compiles the LLEXT export library separately and explicitly appends it at the end of Zephyr link commands.

---

The weak syscall symbols generated by `gen_syscalls.py` are currently compiled in the LLEXT subsystem library, which is then linked among all other Zephyr libraries in an unspecified order. This can cause the weak symbols to override the actual syscall implementations, leading to undefined behaviour.

To fix this, place the `syscall_export_llext.c` file in its own separate non-Zephyr library so that it can be manually appended after all other libraries in the link order.

pillo79 committed:
llext: fix link order for weak syscall symbols

The weak syscall symbols generated by gen_syscalls.py are currently
compiled in the LLEXT subsystem library, which is then linked among all
other Zephyr libraries in an unspecified order. This can cause the weak
symbols to override the actual syscall implementations, leading to
undefined behaviour.

To fix this, the currently generated file is split in two elements:

- syscall_exports_llext.c contains the EXPORT_SYMBOL directives for all
  syscalls. This part can be compiled with the LLEXT library and linked
  among all other Zephyr libraries, and ensures all syscalls symbols
  are preserved by the linker.

- syscall_weakdefs_llext.c contains the weak definitions for all syscalls.
  This file is compiled in a separate library that is linked last, so
  that the weak symbols are only used if no other implementation is
  available.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
@tejlmand @teburd friendly ping. PTAL & thanks for the feedback!

tejlmand commented on CMakeLists.txt:
```diff
@@ -841,15 +841,15 @@ if(CONFIG_LEGACY_GENERATED_INCLUDE_PATH)
   --base-output      include/generated/zephyr/syscalls    # Write to this dir
   --syscall-dispatch include/generated/zephyr/syscall_dispatch.c # Write this file
-  --syscall-export-llext  include/generated/zephyr/syscall_export_llext.c
+  --syscall-export-llext syscall_export_llext.c
```
why is this file no longer placed in `include/generated/zephyr` like the other Zephyr generated file ?

pillo79 replied:
Not a requirement, however I wanted to clearly show that unlike all other files in `include/generated` (AFAICS), this is directly compiled standalone to its own library. Let me know if I should revert this.

tejlmand commented on CMakeLists.txt:
```diff
@@ -971,6 +971,12 @@ endforeach()
+    # This library contains weak symbols for all z_impl_* syscalls, and must be
+    # placed after all the actual syscall implementations in the link order to
+    # prevent the weak symbols from overriding the actual implementations.
+    list(APPEND WHOLE_ARCHIVE_LIBS syscall_export_llext)
```
we should not have this deep knowledge of `syscall_export_llext` library at this location.

In fact, this commit 321e395a8f4e79ea74b305c248c2f5aec96a4202 should never have been approved in the form it was proposed, but seems I was removed as assignee on the PR (and probably also too busy with HWMv2 at the time).

This change exactly shows the problem with having a `WHOLE_ARCHIVE_LIBS` in its current form.

In general only Zephyr libs should go into the whole-archive and vanilla CMake static will be outside the whole-archive and can be ensured to be picked up when `zephyr_interface` is linked against them.
However, in this specific case, because the CMake static libs has extra dependencies, then that is not possible in this case.

Let me take a closer look of both the needed feature in 321e395a8f4e79ea74b305c248c2f5aec96a4202 and the requirement in this PR and see if I can come up with something cleaner.

pillo79 replied:
Both the `WHOLE_ARCHIVE_LIBS` and this PR deal with the issue of allowing extensions to perform syscalls. The most important requirement is that **all _defined_ syscalls need to be preserved** even if unused by the kernel.

We don't strictly need undefined syscalls to have a NULL `_impl` symbol exported, but having everything whole-archive'd implies that _all_ boilerplate is compiled so every symbol must be present at final link time.

Thanks for looking into this, any advice is appreciated.

pillo79 committed:
llext: enlarge test coverage

There is no reason the syscall tests should not be perfomed on
relocatable builds. This made some test runs pass even though
issues were present in those configurations.

Remove the build gate.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
@tejlmand I hope I cracked the nut this time. :crossed_fingers: 

**v2** of this PR splits the `EXPORT_SYMBOL`s from the weak globals. This way:

* like in current `main`, the exports are `#include`d by the `subsys/llext` Zephyr library. Having this among all the Zephyr libs is actually useful to force inclusion of all syscall implementations. 
* the weak symbols are now generated, compiled in a CMake lib, and placed last in the link order (outside `whole-archive`) by the main CMakeLists.txt, so that part is self-contained and there is no cross-directory stuff. 

(Also, while testing this I was wondering why some test configurations always passed - turns out they were skipping syscall tests. I removed the gate because there is no relation between the LLEXT build type and the inclusion of syscall / kernel object tests.)

pillo79 commented:
@tejlmand @teburd weekly ping for review... I hope the latest version can be an acceptable solution.

tejlmand reviewed (approved):
Currently too busy too address https://github.com/zephyrproject-rtos/zephyr/pull/81580#discussion_r1858132170.

However this should not block the improvements proposed in this PR, therefore a followup enhancement issue has been created: https://github.com/zephyrproject-rtos/zephyr/issues/82780

Pr approved.
